includes/utils/class-amp-validation-utils.php

```diff
@@ -483,4 +487,62 @@ public static function add_validation_callback( $sanitizers ) {
 		return $sanitizers;
 	}
 
+	/**
+	 * Registers the post type to store the validation errors.
+	 *
+	 * @return void.
+	 */
+	public static function post_type() {
+		register_post_type(
+			self::POST_TYPE_SLUG,
+			array(
+				'labels'   => array(
+					'name' => _x( 'AMP Validation Errors', 'post type general name', 'amp' ),
+				),
+				'supports' => false,
+			)
+		);
+	}
+
+	/**
+	 * Stores the validation errors.
+	 *
+	 * After the preprocessors have run, this gets the validation response.
+	 * If a plugin output invalid markup, it stores the response in a custom post type.
+	 *
+	 * @return int|void $post_id The post ID of the custom post type used, or void.
+	 */
+	public static function store_validation_errors() {
+		$response = self::get_response();
+		if ( ! $response[ self::ERROR_KEY ] || empty( $response[ self::PLUGINS_INVALID_OUTPUT ] ) ) {
+			return;
+		}
+		$url = isset( $response['url'] ) ? $response['url'] : null;
+		unset( $response['url'] );
+		$encoded_errors = wp_json_encode( $response );
+		$post_name      = md5( $encoded_errors );
+		$post           = get_page_by_path( $post_name, OBJECT, self::POST_TYPE_SLUG );
+		$post_id        = isset( $post->ID ) ? $post->ID : null;
+
+		// If there's no post storing these specific errors, create one.
+		if ( ! isset( $post_id ) ) {
+			$post_id = wp_insert_post( array(
+				'post_type'    => self::POST_TYPE_SLUG,
+				'post_name'    => $post_name,
+				'post_content' => $encoded_errors,
+			) );
+		}
+
+		// Store the URL where the error(s) appeared.
+		$meta = get_post_meta( $post_id, self::URLS_VALIDATION_ERROR, true );
```

Comment:
Instead of `true` to return a single meta and then append to it with `update_post_meta()` this could instead skip straight to doing `add_post_meta( $post_id, self::URLS_VALIDATION_ERROR, $url, true )`.

Reply:
Thanks, [this commit](https://github.com/Automattic/amp-wp/commit/87df66ac81f186c05e2e07ad849f96b86836b142) begins with `add_post_meta()`, and only runs `update_post_meta()` if there was already a value for that key. Is that what you had in mind?